### ISSUE TITLE
version command does not end

### ISSUE THREAD

**Issue by friedger** [Light-Labs/ryder-cli-proto#2]
I have to press Ctrl+C after running `ryder-cli-proto --ryder-port /dev/pts/5 version`

**Comment by MarvinJanssen** [comment]
Which platform are you experiencing this on? Probably an issue with ryderserial.

**Comment by chazkiker2** [comment]
I'm on macOS Big Sur (v 11.2.3) and experienced the same issue when running `ryder-cli-proto --ryder-port /dev/ttys003 version` on iTerm. 

## Executed Command

```shell
> ryder-cli-proto --ryder-port /dev/ttys003 version
```

## Output

```shell
Ryder Prototype CLI 0.0.1.

All commands require the Ryder port to be set. You can do this by
either specifying it with the option --ryder-port or by exporting
the environment variable RYDER_PORT.

Warning: this is prototype software, never use your real seed phrases.

Use --help to see a list of commands.
^C
```

## Debug Output & Additional details 

### Output when run with debug: 

When run with debug, the following output:

```shell
ryderserial attempt open
ryderserial open
queue data for Ryder: 1 byte(s)
send data to Ryder: 1 byte(s)
data from Ryder 0x047279646572000002000105
```

### `argv` after initialization

`console.log({ argv })` before return statement in first middleware ([line 26 src/index.js](https://github.com/Light-Labs/ryder-cli-proto/blob/6cb6cdb24e61b1ff240832ea373011e264e5a374/src/index.js#L26))

```javascript
{
  argv: {
    _: [ 'version' ],
    '$0': '/usr/local/bin/ryder-cli-proto',
    'ryder-port': '/dev/ttys003'
  }
}
```
### `argv` and `info` after processing command

`console.log({ info, argv })` before resolve statement (`resolve(argv)`) in second middleware ([line 54 src/index.js](https://github.com/Light-Labs/ryder-cli-proto/blob/6cb6cdb24e61b1ff240832ea373011e264e5a374/src/index.js#L54))

```javascript
{
  info: 'ryder\x00\x00\x02\x00\x01',
  argv: {
    _: [ 'version' ],
    '$0': '/usr/local/bin/ryder-cli-proto',
    'ryder-port': '/dev/ttys003',
    _ryder_serial: RyderSerial {
      id: 0,
      port: '/dev/ttys003',
      options: [Object],
      closing: false,
      serial: [SerialPort],
      _events: [Object: null prototype],
      _eventsCount: 3,
      [Symbol(train)]: [],
      [Symbol(state)]: 0,
      [Symbol(ready)]: [],
      [Symbol(watchdog)]: Timeout {
        _idleTimeout: 5000,
        _idlePrev: [TimersList],
        _idleNext: [TimersList],
        _idleStart: 227,
        _onTimeout: [Function: bound serial_watchdog],
        _timerArgs: undefined,
        _repeat: null,
        _destroyed: false,
        [Symbol(refed)]: true,
        [Symbol(kHasPrimitive)]: false,
        [Symbol(asyncId)]: 15,
        [Symbol(triggerId)]: 0
      }
    }
  }
}
```

**Comment by chazkiker2** [comment]
@MarvinJanssen @friedger 

Realizing now that this issue is not due to the `version` command specifically, it's because `version` is not a valid command at all (only `--version` is valid).

If you do the same thing with any arbitrary command that is not valid (i.e., `flamingo` or `arbitrary`), you'll get the same issue. 

INVALID command example: `> ryder-cli-proto --ryder-port /dev/ttys003 flamingo`

<details>

<summary>Output with invalid command</summary>

```shell
# executed command
> ryder-cli-proto --ryder-port /dev/ttys003 flamingo

# debug info
ryderserial attempt open
ryderserial open
queue data for Ryder: 1 byte(s) 
send data to Ryder: 1 byte(s) 
data from Ryder 0x047279646572000002000105

# ryder-cli output
Ryder Prototype CLI 0.0.1.

All commands require the Ryder port to be set. You can do this by
either specifying it with the option --ryder-port or by exporting
the environment variable RYDER_PORT.

Warning: this is prototype software, never use your real seed phrases.

Use --help to see a list of commands.
^C
# REQUIRES ^C input from the user to kill -- RyderSerial doesn't `close` after `send`
```

</details>


VALID COMMAND — `> ryder-cli-proto --ryder-port /dev/ttys003 info`

<details>

<summary>Output from valid command</summary>

```shell
# executed command
> ryder-cli-proto --ryder-port /dev/ttys003 info
# debug info
ryderserial attempt open
ryderserial open
queue data for Ryder: 1 byte(s) 
send data to Ryder: 1 byte(s) 
data from Ryder 0x047279646572000002000105
queue data for Ryder: 1 byte(s) 
send data to Ryder: 1 byte(s) 
data from Ryder 0x047279646572000002000105

# ryder-cli-proto output
Initialised Ryder FW version 0.0.2 on /dev/ttys003

# debug output
ryderserial close

# does NOT require ^C input from the user to kill -- we manually `close` RyderSerial at the end of the `src/cmd/info` script
```

</details>

## TL;DR / Why this issue is occurring

So the reason this is hanging up with the `version` command is b/c we actually hit `RyderSerial.send(RyderSerial.COMMAND_INFO)` [right here](https://github.com/Light-Labs/ryder-cli-proto/blob/6cb6cdb24e61b1ff240832ea373011e264e5a374/src/index.js#L48) no matter WHAT the user passes in as a command. 

Since `RyderSerial.send()` does not internally `RyderSerial.close()` afterward, the connection remains open (but unchanging). Our `ryder-cli-proto/src/index.js` never checks the input command to see if it's valid; so when that connection remains open after `RyderSerial.send`, our `ryder-cli-pro` assumes that the `RyderSerial` is doing work of some sort even though the `RyderSerial.send` command has done its job. So we never `RyderSerial.close()`.

## Possible Fixes 

- (in [ryder-cli-proto/src/index.js](https://github.com/Light-Labs/ryder-cli-proto/blob/master/src/index.js)) Add a middleware in the `yargs` chain to check the given command (`argv["$0"]`) against an array of valid commands; if the given command is valid: execute it. If the given command is invalid: print an error message. 
- (in [ryder-cli-proto/src/index.js](https://github.com/Light-Labs/ryder-cli-proto/blob/master/src/index.js)) Add a call to `argv._ryder_serial.close();` somewhere in the async middleware that actually constructs an instance of `RyderSerial`
- (in [ryderserial-proto/src/index.js](https://github.com/Light-Labs/ryderserial-proto/blob/master/src/index.js)) Change the byte returned from the `RyderSerial.send` function to somehow denote that the information sent was not understood as any sort of known command (like `RyderSerial.COMMAND_INFO`).

Whether the `RyderSerial.send(SOME_BYTE)` remains open (i.e., does not call `RyderSerial.close()` by DESIGN is what I am unsure of. If this is an intentional feature of the `ryderserial-proto`, we should certainly implement one of the first two fixes listed above. If this is a bug in `ryderserial-proto` that needs to be fixed, we can implement the third solution over there... but even so it may be worth adding a check for a valid command to avoid unwanted side effects. 

I'll draft a PR (and link this issue) adding a check for the input command on this side of things and y'all can make a decision either way :)

**Comment by MarvinJanssen** [comment]
Interesting! Thank you for investigating all of this.

Yes, `send()` has nothing to do with opening and closing. Because of the async nature, send should never close. If you send a random byte to the Ryder, it will error out with an ERROR_UNKNOWN_COMMAND. The CLI just always sends an INFO first to check if the device on the other side is actually a Ryder. It is up to the individual commands to close the port after (made this way due to the nature of the yargs package). It should probably be refactored so that once the command has finished, the yargs wrapper closes the port.

Honestly, yargs is just kind of crappy which lead to this design. I would be up for redoing the CLI with a different package.

**Comment by chazkiker2** [comment]
That makes a lot of sense. I've got a simple fix in PR #3  ready that at least checks to see that the commands are valid (dynamically with `src/cmds/` so we don't have to hard-code command keywords). If the command is valid, it'll be carried out as usual; otherwise, we'll print an error message and exit the process before attempting to connect to the `RyderSerial` instance.

and it's true that refactoring away from yargs might be a good move. Honestly just wish this was in TS haha. but no biggie either way.

**Comment by MarvinJanssen** [comment]
Totally, a bunch of the code base has come about through experimentation. TS would be very welcome.
